documentation/examples/custom-sd/adapter-usage/main.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"github.com/prometheus/prometheus/documentation/examples/custom-sd/adapter"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+var (
+	a             = kingpin.New("sd adapter usage", "Tool to generate file_sd config files for unimplemented SD mechanisms.")
+	outputFile    = a.Flag("output.file", "Output file for file_sd compatible file. Prints to stdout if none provided.").Default("/tmp/custom_sd.json").String()
+	listenAddress = a.Flag("listen.address", "The address the http sd is listening on for requests.").String()
+	logFile       = a.Flag("log.file", "Location of logfile.").String()
+	logger        log.Logger
+)
+
+type sdConfig struct {
+	listenAddress   string
+	refreshInterval int
+	name            string
+}
+
+type discovery struct {
+	listenAddress   string
+	refreshInterval time.Duration
+	name            string
+}
+
+type targetsResponse struct {
+	services map[string][]string
+}
+
+func (d *discovery) jsonToTargets(data []byte) ([]*targetgroup.Group, error) {
+	var arbJSON interface{}
+	var tgs []*targetgroup.Group
+
+	err := json.Unmarshal(data, &arbJSON)
+	if err != nil {
+		return nil, err
+	}
+
+	targetsMap := arbJSON.(map[string]interface{})
```

Comment:
This will panic on bad data

I'd suggest using normal JSON parsing rather than walking the struct by hand.

Reply:
This is the recommended way to decode arbitrary json: https://blog.golang.org/json-and-go

I just wrote the example HTTP SD to return json like this:
```
{"euw":["database"],"use":["database","webserver"]}
```
Here each field is some arbitrary datacenter name with an array of the service names running there. So if we want to `Unmarshal` into an actual struct and not have a type assertion I'll change the structure of the data that the HTTP SD returns. But that depends on how strongly you feel this needs to change. 

The alternative is that I just change that line to `targetsMap, ok := arbJSON.(map[string]interface{})`. This will remove to possibility of a panic and we can check the value of `ok` and return if we couldn't assert to the right type.

Reply:
All the exporters I know of use standard json unmarshalling, let's not confuse the matter.